src/structure_managers/structure_manager_centers.cc

```diff
@@ -47,8 +48,17 @@ namespace rascal {
 
     // set the references to the particles positions
     for (size_t id{0}; id < this->natoms; ++id) {
-      this->atoms_index[0].push_back(id);
-      this->offsets.push_back(id);
+      if (center_atoms_mask(id)) {
+        this->atoms_index[0].push_back(id);
+        this->offsets.push_back(id);
+      }
+    }
+
+    for (size_t id{0}; id < this->natoms; ++id) {
```

Comment:
Wait, what? So the references to the central atoms still get stored, but now they're just pushed to the end of the array? And iteration over the centers only iterates over then first `n_centers` atoms, implicitly skipping the atoms that were not selected as centers? Where is that mentioned/documented?

Another issue: This changes the ordering of the atoms from what was passed in, potentially breaking user and developer expectations. This definitely needs to be documented prominently.

Reply:
I added some documentation

Reply:
it does not change the ordering of the input atoms, it removes some from the list that will be iterated over with. Is is clearer ?

Reply:
No, not really, because this sequence of for-loops makes it look like you're first iterating over the center atoms, then over the atoms that are not marked as centers, but then doing exactly the same thing with them (adding their IDs to the same two arrays). So this has the effect of reordering the arrays so that the IDs of the centers come first, no? Then where is the actual center selection done? Do you only iterate up to `n_center_atoms` and skip the rest of the array?

Reply:
> So this has the effect of reordering the arrays so that the IDs of the centers come first, no?

the value of id is given by the original order provided by the user, the mask on the center orders these indices differently. This reordering does not matter since atoms that are not centers don't belong to the main loop over the manager: the center atoms will come in the order they are provided by the user.
What can happen is that the order the neighbors are iterated over changes depending on which center is selected. I don't see how it can be an issue since representations should not be sensitive to this arbitrary ordering.

>Then where is the actual center selection done?  Do you only iterate up to n_center_atoms and skip the rest of the array?

it is done here by the magic of the order and the fact that the manager size == n_center_atoms.

Reply:
Right, so the iteration over neighbours is guaranteed to happen in the same order _only_ if the atomic symbols, positions, cell, cutoff, and the center mask are all the same. The reason this might be important is e.g. in indexing dense arrays from a representation manager, you might have some expectation as to how these are ordered -- especially if you're processing these outside of librascal. What happens if you get an (atomic) kernel matrix and you want to know which row corresponds to which atom? In the case of center selection, it looks like the order of the selected centers should be preserved, but this should definitely be written down somewhere.

Reply:
Is this solved?

Reply:
Not sure, I think the implications of the reordering still need to be documented.